Avoid frtree to be written in plots.root + other updates for standalone workflow (samples and only initialStep for BH, STD and CE builds)

mmasciov opened the pull request:
Brings size of plots.root down by ~300x.
Results are identical, e.g., can compare:
(1a) http://uaf-10.t2.ucsd.edu/~mmasciov/MkFit_pixelPairStep/simval_test_iter4/
(2a) http://uaf-10.t2.ucsd.edu/~mmasciov/MkFit_pixelPairStep/SAV_pixelPairStep/SAV_pixelPairStep_iter4/

or 

(1b)  http://uaf-10.t2.ucsd.edu/~mmasciov/MkFit_pixelPairStep/simval_test_iter22/
(2b) http://uaf-10.t2.ucsd.edu/~mmasciov/MkFit_pixelPairStep/SAV_pixelPairStep/SAV_pixelPairStep_iter22/

As for BH, STD and CE builds (not multi-iter), results are consistent after using only initialStep.
E.g.,  http://uaf-10.t2.ucsd.edu/~mmasciov/MkFit_pixelPairStep/SVA_noMultiIter_testV3/ to be compared to http://uaf-10.t2.ucsd.edu/~mmasciov/MkFit_pixelPairStep/SVA_multiIter/simvalseed_test_iter4/

Note (1): samples are latest available, including fix in PR #330.
Note (2): for non-multi-iter builds (BH, STD and CE), this limits validation to initialStep. Potential extension of BH and STD to multi-iter is not addressed in this PR.
Note (3): few of the scripts (not of common use anymore) still refer to SNB architecture, or are labeled as "SNB". As phi1 (SNB) is still online, this has not been changed in this PR.

mmasciov committed:
Avoid frtree to be written in plots.root

mmasciov commented:
Before merging, I will push extra changes to point to most recent samples everywhere in validation scripts.

mmasciov committed:
Updating xeon_scripts to latest (2021) samples

mmasciov committed:
Updating val_scripts to latest (2021) samples

mmasciov commented:
The PR is ready for review/merge on my side.

mmasciov committed:
Adding argument to move-benchmarks to collect either standard benchmarks, or multi-iter ones

slava77 commented on val_scripts/validation-cmssw-benchmarks.sh:
```diff
@@ -32,6 +32,12 @@ case ${inputBin} in
         subdir=
         file=pu50-ccc-hs.bin
         ;;
+"112X_TTbar_PU50_MULTI")
```
do we get meaningful results with the multi-iter bin files for this non-multi benchmark setup?

mmasciov replied:
I don't see why not, but let me test this.

slava77 replied:
is the code in this case selecting iter=0/initialStep or does it inclusively run on all seeds and hits?

mmasciov replied:
I believe it uses everything. @osschar should correct me.
In fact, we had talked of replacing CE building with MultiIter building running only for one iteration in the past, but I don't think this was done so far.

mmasciov replied:
OK, I will push another commit to do what I described above. Right now, with a partial change:
http://uaf-10.t2.ucsd.edu/~mmasciov/MkFit_pixelPairStep/simval_test_CE_noMultiIter/efficiency/

Efficiency for CE (or STD) is the same as expected (http://uaf-10.t2.ucsd.edu/~mmasciov/MkFit_pixelPairStep/SAV_pixelPairStep/SAV_pixelPairStep_iter4/efficiency/?match=); on the other hand, CMSSW is running over all seeds/hits. Will find a way to solve it, and then everything should be consistent.

osschar replied:
Yes, I don't think we do any seed filtering in the "old" bh/std/ce building functions. We need to decide how to rewrite them.

mmasciov replied:
With the following commits, I'm solving this discrepancy:
http://uaf-10.t2.ucsd.edu/~mmasciov/MkFit_pixelPairStep/SVA_noMultiIter_testV3/
(to be compared, e.g., to http://uaf-10.t2.ucsd.edu/~mmasciov/MkFit_pixelPairStep/SVA_multiIter/simvalseed_test_iter4/, which matches the above multi-iter example, but run on 500 evts as opposed to 1000, for consistency)

Let me note that builders have been updated to only use initialStep information, for consistency with the past.

mmasciov committed:
Reducing # events to 250 to avoid (memory-related) crash in 2021 TTbar

mmasciov committed:
Collect build plots by default

mmasciov committed:
Update builders other than multi-iter to allow to run only on initialStep by default

mmasciov committed:
Update validation plotting to be consistent both with and without multi-iter

mmasciov committed:
Updating benchmark val scripts to be consistent both with and without multi-iter

mmasciov committed:
Removing BH from default 'forPR' val builds

slava77 commented:
I noticed that validation-cmssw-benchmarks-multiiter.sh works only with forConf.
Does this PR allow to expand to forPR? or is the idea to reduce STD and BH to be only in the initialStep?

mmasciov commented:
This only reduces STD and BH to initialStep, while we use MultiIter (an extension of CE) for multiple iterations. 
I would go with this first. Then, we can think of extending STD and BH to other iterations, but so far this has never been tested, and would require non-negligible testing/changes.
Also, I'm personally not sure it makes sense to keep using STD and BH if we are not going to use any of the two.

mmasciov commented:
This PR is ready to be reviewed/merged. No additional change appears to be required in order to limit validation of builds other than multi-iter to initialStep-only, which is the goal of this PR, together with updating to latest samples and reduce size of validation output by avoiding un-necessary copies of trees.

slava77 commented on val_scripts/validation-cmssw-10mu-fulldet-build-extrectracks.sh:
```diff
@@ -2,50 +2,30 @@
-ECP2=${dir}/10muEta17to24Pt1to10/${file}
+dir=/data2/slava77/samples/2021/
+subdir=10muPt0p2to10HS/
+file=memoryFile.fv5.default.210623-b62fc88.bin
```
```suggestion
file=memoryFile.fv5.default.210703-d239b45.bin
```

slava77 commented on val_scripts/validation-cmssw-10mu-fulldet-build.sh:
```diff
@@ -2,30 +2,23 @@
-ECP2=${dir}/10muEta17to24Pt1to10/${file}
+dir=/data2/slava77/samples/2021/
+subdir=10muPt0p2to10HS/
+file=memoryFile.fv5.default.210623-b62fc88.bin
```
```suggestion
file=memoryFile.fv5.default.210703-d239b45.bin
```

slava77 commented on val_scripts/validation-cmssw-benchmarks-multiiter.sh:
```diff
@@ -26,6 +26,12 @@ case ${inputBin} in
+        echo "Inputs from 2021 TTbar (PU50) sample with multiple iterations and hit binary mask"
+        dir=/data2/slava77/samples/
+        subdir=2021/11834.0_TTbar_14TeV+2021/AVE_50_BX01_25ns/
+        file=memoryFile.fv5.default.210623-b62fc88.bin
```
```suggestion
        file=memoryFile.fv5.default.210703-d239b45.bin
```

slava77 commented on val_scripts/validation-cmssw-benchmarks.sh:
```diff
@@ -32,6 +32,12 @@ case ${inputBin} in
+        echo "Inputs from 2021 TTbar (PU50) sample with multiple iterations and hit binary mask"
+        dir=/data2/slava77/samples/
+        subdir=2021/11834.0_TTbar_14TeV+2021/AVE_50_BX01_25ns/
+        file=memoryFile.fv5.default.210623-b62fc88.bin
```
```suggestion
        file=memoryFile.fv5.default.210703-d239b45.bin
```

slava77 commented on val_scripts/validation-cmssw-benchmarks.sh:
```diff
@@ -40,6 +46,14 @@ case ${inputBin} in
+        echo "Inputs from 2021 10mu sample with multiple iterations and hit binary mask"
+        dir=/data2/slava77/samples
+        subdir=2021/10muPt0p2to1000HS
+        file=memoryFile.fv5.default.210623-b62fc88.bin
```
```suggestion
        file=memoryFile.fv5.default.210703-d239b45.bin
```

slava77 commented on val_scripts/validation-cmssw-ttbar-fulldet-build-extrectracks.sh:
```diff
@@ -2,18 +2,19 @@
-dir=/data2/slava77/samples/2017/pass-4874f28/initialStep
-file=memoryFile.fv3.clean.writeAll.recT.072617.bin
+dir=/data2/slava77/samples/2021/11834.0_TTbar_14TeV+2021/
+file=memoryFile.fv5.default.210623-b62fc88.bin
```
```suggestion
file=memoryFile.fv5.default.210703-d239b45.bin
```

slava77 commented on val_scripts/validation-cmssw-ttbar-fulldet-build.sh:
```diff
@@ -2,21 +2,22 @@
-dir=/data2/slava77/samples/2017/pass-4874f28/initialStep
-file=memoryFile.fv3.clean.writeAll.recT.072617.bin
+dir=/data2/slava77/samples/2021/11834.0_TTbar_14TeV+2021/
+file=memoryFile.fv5.default.210623-b62fc88.bin
```
```suggestion
file=memoryFile.fv5.default.210703-d239b45.bin
```

slava77 commented on val_scripts/validationMIC-build-10mu.sh:
```diff
@@ -1,55 +1,33 @@
-ECP1=${BIN_DATA_PATH}/10muEta055to175Pt1to10/memoryFile.fv3.recT.072617.bin
-ECP2=${BIN_DATA_PATH}/10muEta17to24Pt1to10/memoryFile.fv3.recT.072617.bin
+[ -e "$BIN_DATA_PATH" ] || BIN_DATA_PATH=/data2/slava77/samples/2021/
+fin10mu=${BIN_DATA_PATH}/10muPt0p2to10HS/memoryFile.fv5.default.210623-b62fc88.bin
```
```suggestion
fin10mu=${BIN_DATA_PATH}/10muPt0p2to10HS/memoryFile.fv5.default.210703-d239b45.bin

slava77 commented on val_scripts/validationMIC-build-PU70.sh:
```diff
@@ -1,15 +1,15 @@
-[ -e "$BIN_DATA_PATH" ] || BIN_DATA_PATH=/store/disk00/slava77/analysis/CMSSW_9_1_0_pre1-tkNtuple/run1000/2017/pass-4874f28/initialStep
-fin=${BIN_DATA_PATH}/PU70/10224.0_TTbar_13+TTbar_13TeV_TuneCUETP8M1_2017PU_GenSimFullINPUT+DigiFullPU_2017PU+RecoFullPU_2017PU+HARVESTFullPU_2017PU/memoryFile.fv3.clean.writeAll.recT.072617.bin
+[ -e "$BIN_DATA_PATH" ] || BIN_DATA_PATH=/data2/slava77/samples/2021/11834.0_TTbar_14TeV+2021/
+fin=${BIN_DATA_PATH}/AVE_70_BX01_25ns/memoryFile.fv5.default.210623-b62fc88.bin
```
```suggestion
fin=${BIN_DATA_PATH}/AVE_70_BX01_25ns/memoryFile.fv5.default.210703-d239b45.bin
```

slava77 commented on xeon_scripts/benchmark-cmssw-ttbar-fulldet-build.sh:
```diff
@@ -69,8 +64,9 @@ else
-file=pu50-ccc-hs.bin
+dir=/data2/slava77/samples/
+subdir=2021/11834.0_TTbar_14TeV+2021/AVE_50_BX01_25ns/
+file=memoryFile.fv5.default.210623-b62fc88.bin
```
```suggestion
file=memoryFile.fv5.default.210703-d239b45.bin
```

slava77 commented on xeon_scripts/benchmarkMIC-build.sh:
```diff
@@ -1,7 +1,7 @@
-[ -e "$BIN_DATA_PATH" ] || BIN_DATA_PATH=/store/disk00/slava77/analysis/CMSSW_9_1_0_pre1-tkNtuple/run1000/2017/pass-4874f28/initialStep
-fin=${BIN_DATA_PATH}/PU70/10224.0_TTbar_13+TTbar_13TeV_TuneCUETP8M1_2017PU_GenSimFullINPUT+DigiFullPU_2017PU+RecoFullPU_2017PU+HARVESTFullPU_2017PU/memoryFile.fv3.clean.writeAll.recT.072617.bin
+[ -e "$BIN_DATA_PATH" ] || BIN_DATA_PATH=/data2/slava77/samples/2021/11834.0_TTbar_14TeV+2021/
+fin=${BIN_DATA_PATH}/AVE_70_BX01_25ns/memoryFile.fv5.default.210623-b62fc88.bin
```
```suggestion
fin=${BIN_DATA_PATH}/AVE_70_BX01_25ns/memoryFile.fv5.default.210703-d239b45.bin
```

slava77 commented on xeon_scripts/benchmarkMIC-build.sh:
```diff
@@ -49,26 +49,19 @@ runBenchmark 1
 make clean
 make distclean
 
+fin10mu=/data2/slava77/samples/2021/10muPt0p2to10HS/memoryFile.fv5.default.210623-b62fc88.bin 
```
```suggestion
fin10mu=/data2/slava77/samples/2021/10muPt0p2to10HS/memoryFile.fv5.default.210703-d239b45.bin
```

slava77 commented on xeon_scripts/debug-test.sh:
```diff
@@ -8,9 +8,9 @@ source xeon_scripts/init-env.sh
-file=memoryFile.fv3.clean.writeAll.recT.072617.bin
+dir=/data2/slava77/samples/2021/
+subdir=10muPt0p2to10HS
+file=memoryFile.fv5.default.210623-b62fc88.bin
```
```suggestion
file=memoryFile.fv5.default.210703-d239b45.bin
```

slava77 commented on xeon_scripts/stress-test-main.sh:
```diff
@@ -71,8 +68,9 @@ else
-file=memoryFile.fv3.clean.writeAll.CCC1620.recT.082418-25daeda.bin
+dir=/data2/slava77/samples/
+subdir=2021/11834.0_TTbar_14TeV+2021/AVE_50_BX01_25ns/
+file=memoryFile.fv5.default.210623-b62fc88.bin
```
```suggestion
file=memoryFile.fv5.default.210703-d239b45.bin
```

slava77 commented on xeon_scripts/throughput-test-main.sh:
```diff
@@ -40,10 +36,11 @@ fi
-file=memoryFile.fv3.clean.writeAll.CCC1620.recT.082418-25daeda.bin
+dir=/data2/slava77/samples/
+subdir=2021/11834.0_TTbar_14TeV+2021/AVE_50_BX01_25ns/
+file=memoryFile.fv5.default.210623-b62fc88.bin
```
```suggestion
file=memoryFile.fv5.default.210703-d239b45.bin
```

slava77 reviewed (commented):
the files are now on phi3

mmasciov committed:
Replacing samples in all val and xeon scripts with latest (with fix)

mmasciov commented:
All samples have been replaced with the latest by @slava77 (with fix in PR #330).
PR should be now ready for merge.

slava77 commented on mkFit/buildtestMPlex.cc:
```diff
@@ -117,17 +117,53 @@ namespace
+    TrackVec seeds1;   
+    unsigned int algorithms[]={ 4 }; //only initialStep
+    
+    for (auto &s : ev.seedTracks_)
```
```suggestion
    for (auto const&s : ev.seedTracks_)
```

slava77 commented on mkFit/buildtestMPlex.cc:
```diff
@@ -117,17 +117,53 @@ namespace
+    
+    builder.begin_event(&job, &ev, __func__);
+    
+    if (Config::sim_val_for_cmssw) {
+      builder.root_val_dumb_cmssw();
+    }
+    
+    builder.end_event();
+  } 
+  else{
+    MkJob job( { Config::TrkInfo, itconf, eoh, nullptr } );
+  
+    builder.begin_event(&job, &ev, __func__);
+    
+    if (Config::sim_val_for_cmssw) {
+      builder.root_val_dumb_cmssw();
+    }
+    
+    builder.end_event();
   }
```
```suggestion
  } 
  else{
    MkJob job( { Config::TrkInfo, itconf, eoh, nullptr } );
  
    builder.begin_event(&job, &ev, __func__);
  }

  if (Config::sim_val_for_cmssw) {
    builder.root_val_dumb_cmssw();
  }
  builder.end_event();
```

slava77 commented on mkFit/buildtestMPlex.cc:
```diff
@@ -136,7 +172,34 @@ void runBuildingTestPlexDumbCMSSW(Event& ev, const EventOfHits &eoh, MkBuilder&
+    
+    unsigned int algorithms[]={ 4 }; //only initialStep
+    
+    for (auto &s : ev.seedTracks_)
```
```suggestion
    for (auto const&s : ev.seedTracks_)
```

slava77 commented on mkFit/buildtestMPlex.cc:
```diff
@@ -200,7 +263,34 @@ double runBuildingTestPlexBestHit(Event& ev, const EventOfHits &eoh, MkBuilder&
+    
+    unsigned int algorithms[]={ 4 }; //only initialStep
+    
+    for (auto &s : ev.seedTracks_)
```
```suggestion
    for (auto const&s : ev.seedTracks_)
```

slava77 commented on mkFit/buildtestMPlex.cc:
```diff
@@ -263,7 +353,34 @@ double runBuildingTestPlexStandard(Event& ev, const EventOfHits &eoh, MkBuilder&
+    
+    unsigned int algorithms[]={ 4 }; //only initialStep
+
+    for (auto &s : ev.seedTracks_)
```
```suggestion
    for (auto const&s : ev.seedTracks_)
```

slava77 commented on xeon_scripts/benchmarkMIC-build.sh:
```diff
@@ -1,7 +1,7 @@
-fin=${BIN_DATA_PATH}/PU70/10224.0_TTbar_13+TTbar_13TeV_TuneCUETP8M1_2017PU_GenSimFullINPUT+DigiFullPU_2017PU+RecoFullPU_2017PU+HARVESTFullPU_2017PU/memoryFile.fv3.clean.writeAll.recT.072617.bin
+[ -e "$BIN_DATA_PATH" ] || BIN_DATA_PATH=/data2/slava77/samples/2021/11834.0_TTbar_14TeV+2021/
+fin=${BIN_DATA_PATH}/AVE_70_BX01_25ns/memoryFile.fv5.default.210703-d239b45.bin
 
```
this file has `SNB_` in the results prefix.
Do we even need it now ? (otoh, phi1 is still alive, so, perhaps OK)

mmasciov replied:
I am unsure about the current usage of this script, but as you notes phi1 is still there. So, I left it unchanged for the time being.

mmasciov replied:
Have added a note to the main description that this is not addressed in this PR, despite those scripts not being of common use anymore. Resolving this, since it'll be mentioned above for history.

mmasciov committed:
Add const to loop over seed tracks in buildtestMPlex

leonardogiannini reviewed (commented):
I don't understand this part https://github.com/trackreco/mkFit/pull/328/files/9d9a1beab800d77565f308f2f05e9c7b5008d1a7#diff-9bd117b4474e1261389e0f33d9f898dc985c780ab0e7c9c12339ce68867a0e66R125, which is already in MkBuilder. I also seemed to have problems with selection before relabeling here https://github.com/trackreco/mkFit/blob/devel/mkFit/MkBuilder.cc#L936-L937 if the order of our iterations is not the same as cmssw.

mmasciov committed:
Revert change in CMSSW dumb build and limit CMSSW val to first iteration in non-multi-iter val script

mmasciov commented:
> I don't understand this part https://github.com/trackreco/mkFit/pull/328/files/9d9a1beab800d77565f308f2f05e9c7b5008d1a7#diff-9bd117b4474e1261389e0f33d9f898dc985c780ab0e7c9c12339ce68867a0e66R125, which is already in MkBuilder. I also seemed to have problems with selection before relabeling here https://github.com/trackreco/mkFit/blob/devel/mkFit/MkBuilder.cc#L936-L937 if the order of our iterations is not the same as cmssw.

That part was aimed at reducing the size of the CMSSW valtree.
However, as you noted, the same could be achieved by requiring "--num-iters-cmssw 1" in the validation script, for CMSSW.
Just for confirmation, results are unchanged while the size of the CMSSW valtree is reduced by a factor of ~6x:
- CMSSW validation with original change (vs. original): http://uaf-10.t2.ucsd.edu/~mmasciov/MkFit_pixelPairStep/valCMSSW_test/simvalCMSSW_asisVSoriginal/efficiency/
- CMSSW validation with "--num-iters-cmssw 1" (vs. original): http://uaf-10.t2.ucsd.edu/~mmasciov/MkFit_pixelPairStep/valCMSSW_test/simvalseedCMSSW_numItersVSoriginal/efficiency/

Hence, I reverted my change in the last commit, and just added "--num-iters-cmssw 1" in the non-multi-iter CMSSW validation script, for the CMSSW validation command.
